src/codeGenerator/evaluator.ts

```diff
@@ -353,14 +406,143 @@ export class Evaluator implements ExpressionVisitor {
+        } else if (trigger instanceof CombinedTrigger) {
+            results = results.concat(this.parseTrigger(trigger.lhs, env, sourceExpression));
+            results = results.concat(this.parseTrigger(trigger.rhs, env, sourceExpression));
+        }
+        return results;
     }
-    visitFunction(_astNode: MCFunction, _env: EvaluatorEnv) : EvaluatorData {
-        throw new Error("Method not implemented.");
+
+    // FIXME: lots of switch on type nonsense
+    parseCommands(astCommands: Command[], env: EvaluatorEnv, sourceExpression: Expression) : string[] {
+        let commands: string[] = [];
+        for (let astCommand of astCommands) {
+            if (astCommand instanceof Grant) {
+                commands.push(`advancement grant @p only ${this.evaluateExpectType(astCommand.name, env, "string", "grant command parameter")}`);
+            } else if (astCommand instanceof Revoke) {
+                commands.push(`advancement revoke @p only ${this.evaluateExpectType(astCommand.name, env, "string", "revoke command parameter")}`);
+            } else if (astCommand instanceof Execute) {
+                commands.push( `function ${this.evaluateExpectType(astCommand.name, env, "string", "execute command parameter")}`);
+            } else if (astCommand instanceof RawCommand) {
+                let result = astCommand.command.accept(this, env);
+                if (typeof result === "string") {
+                    commands.push(result)
+                } else if (Array.isArray(result)) {
+                    for (let command of result) {
+                        if (typeof command === "string") {
+                            commands.push(command);
+                        } else {
+                            throw new DSLTypeError(sourceExpression, `expected string array in raw command list, got item of type ${typeof command}`);
+                        }
+                    }
+                } else {
+                    throw new DSLTypeError(sourceExpression, `expected string in raw command list, got item of type ${typeof result}`)
+                }
+            } else {
+                throw new DSLEvaluationError(sourceExpression, `on: Unknown command type ${astCommand.constructor.name}`);
+            }
+        }
+        return commands;
+    }
+
+    visitOn(astNode: On, env: EvaluatorEnv) : EvaluatorData {
+        let commands = this.parseCommands(astNode.commands, env, astNode);
+
+        // Read the advancement trigger and prepare a Minecraft function
+        let fnValue: Store.FunctionValue;
+        let fnName = this.genFunctionName(astNode.trigger.constructor.name);
+        let advName = ""; // FIXME: no advancement name to return for load and tick
+        if (astNode.trigger instanceof Load) {
+            fnValue = Store.FunctionValue.onLoad(fnName, commands);
+        } else if (astNode.trigger instanceof Tick) {
+            fnValue = Store.FunctionValue.onTick(fnName, commands);
```

Comment:
hmm we could create an Advancement for this? https://minecraft.fandom.com/wiki/Advancement/JSON_format#minecraft:tick

or would we rather it be done as a Tag?

Reply:
Will work on this in a follow up - I think using an advancement here is more future proof since it supports predicates & more complex trigger conditions.